### PR TITLE
feature: add modify tg command. fix wdrift ma length

### PR DESCRIPTION
**Pull request by zenixls2**
add `/modify` command to modify strategy fields that is attributed with both `json` and `modifiable:"true"`.  
also will show all the modifiable keys as button in tg.  
FYI.
to pass strings to the `interact`, users need to type `\"string\"` in tg.

### CODE REVIEW

**Comment by zenixls2** [comment]
`/save` command is still under design. only field values in the config file are modifiable.
comments should be kept as before.

File: pkg/bbgo/config.go [hunk 1]
```diff
@@ -478,6 +480,10 @@ func Load(configFile string, loadStrategies bool) (*Config, error) {
 	if err := yaml.Unmarshal(content, &config); err != nil {
 		return nil, err
 	}
+	config.ConfigFile = configFile
+	for _, session := range config.Sessions {
+		session.ConfigPath = config.ConfigFile
+	}
```
Comment (c9s) [inline pkg/bbgo/config.go:486]: Can you pull out the config things to another PR? if you would like this PR to get merged first.

I think the config saving stuff needs to be designed carefully.

File: pkg/bbgo/modify.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+		val = val.Elem()
+	}
+	var targetName string
+	var currVal interface{}
+	var mapping map[string]string
+	// currently we only allow users to modify the first layer of fields
+	RegisterCommand("/modify", "Modify config", func(reply interact.Reply) {
+		reply.Message("Please choose the field name in config to modify:")
+		mapping = make(map[string]string)
+		for i := 0; i < val.Type().NumField(); i++ {
```
Comment (c9s) [inline pkg/bbgo/modify.go:26]: can you pull out this to the dynamic.GetModifiableFields or dynamic.GetChangableFields function?

File: pkg/bbgo/modify.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
+package bbgo
```
Comment (c9s) [inline pkg/bbgo/modify.go:1]: will be better to put this in the file `interact_modify`

File: pkg/bbgo/modify.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+func RegisterModifier(s interface{}) {
+	val := reflect.ValueOf(s)
+	if val.Type().Kind() == util.Pointer {
+		val = val.Elem()
+	}
+	var targetName string
+	var currVal interface{}
+	var mapping map[string]string
+	// currently we only allow users to modify the first layer of fields
+	RegisterCommand("/modify", "Modify config", func(reply interact.Reply) {
```
Comment (c9s) [inline pkg/bbgo/modify.go:23]: wondering if we should use `change` instead of `modify`?

File: pkg/bbgo/session.go [hunk 5]
```diff
@@ -40,6 +40,7 @@ type ExchangeSession struct {
 	Secret       string             `json:"secret,omitempty" yaml:"secret,omitempty"`
 	Passphrase   string             `json:"passphrase,omitempty" yaml:"passphrase,omitempty"`
 	SubAccount   string             `json:"subAccount,omitempty" yaml:"subAccount,omitempty"`
+	ConfigPath   string
```
Comment (c9s) [inline pkg/bbgo/session.go:43]: ditto, Can you pull out the config things to another PR? if you would like this PR to get merged first.

I think the config saving stuff needs to be designed carefully.

File: pkg/strategy/drift/driftma.go [hunk 6]
```diff
@@ -14,7 +14,13 @@ type DriftMA struct {
 
 func (s *DriftMA) Update(value, weight float64) {
 	s.ma1.Update(value)
+	if s.ma1.Length() == 0 {
+		return
+	}
 	s.drift.Update(s.ma1.Last(), weight)
+	if s.drift.Length() == 0 {
+		return
+	}
```
Comment (c9s) [inline pkg/strategy/drift/driftma.go:23]: would be great if this can be a single PR, this looks like a fix?

File: pkg/bbgo/modify.go (hunk 2 above)
Comment (zenixls2) [reply pkg/bbgo/modify.go]: done

File: pkg/bbgo/config.go (hunk 1 above)
Comment (zenixls2) [reply pkg/bbgo/config.go]: done

File: pkg/bbgo/session.go (hunk 5 above)
Comment (zenixls2) [reply pkg/bbgo/session.go]: done

File: pkg/strategy/drift/driftma.go (hunk 6 above)
Comment (zenixls2) [reply pkg/strategy/drift/driftma.go:23]: I would like to merge this fix together

File: pkg/bbgo/modify.go (hunk 4 above)
Comment (zenixls2) [reply pkg/bbgo/modify.go]: maybe still modify?
notice that we also have the cmd 'modifyposition'

File: pkg/bbgo/modify.go (hunk 3 above)
Comment (zenixls2) [reply pkg/bbgo/modify.go]: done

File: pkg/bbgo/interact_modify.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+			reply.Message(fmt.Sprintf("target %s is not modifiable", targetName))
+			return
+		}
+		x := reflect.New(field.Type())
+		xi := x.Interface()
+		if err := json.Unmarshal([]byte(value), &xi); err != nil {
+			reply.Message(fmt.Sprintf("fail to unmarshal the value: %s, err: %v", value, err))
+			return
+		}
+		field.Set(x.Elem())
```
Comment (c9s) [inline pkg/bbgo/interact_modify.go:59]: Need to take care about the Pointer field and Value field, here it looks like only for pointer field? can we add a function for this set operation and add a test for it?

File: pkg/bbgo/interact_modify.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+		newVal := field.Interface()
+		if e, err := json.Marshal(value); err == nil {
+			newVal = string(e)
+		}
+		reply.Message(fmt.Sprintf("update to %v successfully", newVal))
+	})
+
+	RegisterCommand("/save", "Save config", func(reply interact.Reply) {
```
Comment (c9s) [inline pkg/bbgo/interact_modify.go:67]: let's move this save config thing to the next PR

File: pkg/dynamic/field.go [hunk 9]
```diff
@@ -24,3 +27,39 @@ func LookupSymbolField(rs reflect.Value) (string, bool) {
...
+		callback(name, t.Name)
+	}
+}
+
+var ZeroValue reflect.Value = reflect.Zero(reflect.TypeOf(0))
+
+func GetModifiableField(val reflect.Value, name string) (reflect.Value, bool) {
+	field, ok := val.Type().FieldByName(name)
+	if !ok {
+		return ZeroValue, ok
```
Comment (c9s) [inline pkg/dynamic/field.go:55]: it's better to create the zero reflect value from it's field type instead of `0`

File: pkg/strategy/drift/strategy.go [hunk 10]
```diff
@@ -572,35 +560,39 @@ func (s *Strategy) klineHandler1m(ctx context.Context, kline types.KLine) {
...
 	numPending := 0
+	var err error
 	if numPending, err = s.smartCancel(ctx, pricef, atr); err != nil {
 		log.WithError(err).Errorf("cannot cancel orders")
+		s.positionLock.Unlock()
 		return
 	}
 	if numPending > 0 {
-		log.Infof("pending orders: %d, exit", numPending)
+		s.positionLock.Unlock()
```
Comment (c9s) [inline pkg/strategy/drift/strategy.go:587]: types.Position has it's lock, maybe you can reuse it?

File: pkg/bbgo/interact_modify.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+func RegisterModifier(s interface{}) {
+	val := reflect.ValueOf(s)
+	if val.Type().Kind() == util.Pointer {
+		val = val.Elem()
+	}
+	var targetName string
+	var currVal interface{}
+	var mapping map[string]string
+	// currently we only allow users to modify the first layer of fields
+	RegisterCommand("/modify", "Modify config", func(reply interact.Reply) {
```
Comment (c9s) [inline pkg/bbgo/interact_modify.go:23]: You need to let the user to choose the strategy first

File: pkg/strategy/drift/strategy.go [hunk 12]
```diff
@@ -980,6 +977,8 @@ func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, se
 		reply.Message(buffer.String())
 	})
 
+	bbgo.RegisterModifier(s)
```
Comment (c9s) [inline pkg/strategy/drift/strategy.go:980]: multiple strategy instances will override the previous command registration.

File: pkg/bbgo/interact_modify.go (hunk 7 above)
Comment (zenixls2) [reply pkg/bbgo/interact_modify.go:59]: since x is created from reflect.New, as https://pkg.go.dev/reflect#New mentioned, the created `reflect.Value` must be a Pointer. That's why we only care about the pointer type here.

Comment (zenixls2) [reply pkg/bbgo/interact_modify.go:59]: also added test cases

File: pkg/dynamic/field.go (hunk 9 above)
Comment (zenixls2) [reply pkg/dynamic/field.go]: non-existing field will not have type. any error will be presented in ok (the second return value)

File: pkg/strategy/drift/strategy.go (hunk 12 above)
Comment (zenixls2) [reply pkg/strategy/drift/strategy.go:980]: right now only two options:
1. tg vs strategy = 1 to 1 mapping
2. add this to coreInteract

since we currently have only drift applying this configuration, before any useful mechanism been discussed, I prefer to  keep it simple here and recommend users to try 1.

File: pkg/strategy/drift/strategy.go (hunk 10 above)
Comment (zenixls2) [reply pkg/strategy/drift/strategy.go:587]: I believe it's for different purpose.  
this position lock is mainly for price updates (buyPrice, sellPrice, lowestPrice, highestPrice), so should be fine.

File: pkg/bbgo/interact_modify.go (hunk 8 above)
Comment (zenixls2) [reply pkg/bbgo/interact_modify.go]: removed

**Review by c9s — APPROVED** [review]
